scripts/shtf.py

```diff
@@ -0,0 +1,629 @@
+import json
+import logging
+import os
+import time
+from collections import defaultdict
+from dataclasses import dataclass, field
+from datetime import datetime
+from pathlib import Path
+from typing import Callable, DefaultDict, Optional, TypedDict, cast
+
+import click
+import requests
+import toml
+from eth_typing import BlockNumber
+from typing_extensions import NotRequired
+from web3 import Web3
+from web3.constants import ADDRESS_ZERO
+from web3.middleware import geth_poa_middleware
+from web3.types import BlockData
+
+import beamer.contracts
+import beamer.events
+from beamer.config import _merge_dicts
+from beamer.events import ClaimMade, DepositWithdrawn, RequestCreated, RequestFilled
+from beamer.typing import ChainId
+from beamer.util import load_ERC20_abi, setup_logging
+
+logging.disable()
+
+
+class MessageTypes:
+    REQUEST_EXPIRED = "RequestExpired"
+    DEPOSIT_WITHDRAWN = "DepositWithdrawn"
+    UNCLAIMED_FILL = "UnclaimedFill"
+    CHALLENGE_GAME = "ChallengeGame"
+    CHALLENGE_GAME_CLAIM_BY_SOMEONE_ELSE = "ChallengeGameClaimedBySomeoneElse"
+
+
+class MessageType(TypedDict):
+    text: str
+    message_link: NotRequired[str]
+
+
+TokensConfig = dict[str, list[tuple[str, str]]]
+
+
+class ConfigType(TypedDict):
+    agent_address: str
+    deployment_dir: Path
+    notification_system: str
+    rpcs: dict[int, tuple[str, BlockNumber]]
+    explorers: dict[int, str]
+    notification: dict[str, dict[str, str]]
+    tokens: TokensConfig
+
+
+class TransferType(TypedDict):
+    created: NotRequired[RequestCreated]
+    filled: NotRequired[RequestFilled]
+    claimed: NotRequired[dict[int, list[ClaimMade]]]
+    withdrawn: NotRequired[DepositWithdrawn]
+
+
+transfersType = DefaultDict[str, TransferType]
+
+ChainToEvent = dict[ChainId, list[beamer.events.Event]]
+
+
+class MetaType(TypedDict):
+    request_id: str
+    message_type: str
+    message_link: str
+
+
+class NotificationType(TypedDict):
+    meta: MetaType
+    body: str
+
+
+@dataclass
+class Stats:
+    fills: int = 0
+    requests: int = 0
+    withdraws: int = 0
+    volume: int = 0
+
+
+@dataclass
+class TransferStats:
+    volume: int = 0
+    fills: int = 0
+    requests: int = 0
+    withdrawals: int = 0
+    claims: int = 0
+    expired_requests: int = 0
+    expired_requests_volume: int = 0
+
+
+class NotificationState:
+    def __init__(self) -> None:
+        with open(os.path.join(os.path.dirname(__file__), "breadcrumbs.json"), "r") as f:
+            self.notified_state = json.load(f)
+
+    def is_set(self, request_id: str, notified_on: str) -> bool:
+        breadcrumb = self.notified_state.get(str(request_id))
+        if breadcrumb is None:
+            return False
+        else:
+            return breadcrumb.get(notified_on, False)
```

Comment:
just to correct the return type you can use bool(breadcrumb.get..) as if this returns normally then it is probably returning datetime or str?